<div style="align: center; margin: 0; padding: 0; height: 250px;">
    <br>
    <img src="https://storage.googleapis.com/kaggle-datasets-images/3574954/6224332/912c27d4f5b7f3fa6141b2f7adec3cee/dataset-cover.jpg" style="display:block; margin:auto; width:65%; height:100%;">
</div><br><br>

<div style="letter-spacing:normal; opacity:1.;">
<!--   https://xkcd.com/color/rgb/   -->
  <p style="text-align:center; background-color: lightsalmon; color: Jaguar; border-radius:10px; font-family:monospace; 
            line-height:1.4; font-size:32px; font-weight:bold; text-transform: uppercase; padding: 9px;">
            <strong>Bank Credit Scoring Dataset</strong></p>  
            
  <p style="text-align:center; background-color:romance; color: Jaguar; border-radius:10px; font-family:monospace; 
            line-height:1.4; font-size:22px; font-weight:normal; text-transform: capitalize; padding: 5px;"
     >Machine Learning Engineer Module-4:<br>Bank Credit Scoring Approval Estimate Model Deployment<br><br></p>
    
  <div style="align: center;">
  <table style="text-align: center; background-color: romance; color: Jaguar; border-radius: 10px; font-family: monospace;
                  line-height:1.4; font-size: 21px; font-weight: normal; text-transform: capitalize; padding: 5px; 
                  margin: 0 auto;">
    <tr><td style="text-align: left; padding-left: 0px;"
            > Virtual Enviroment with Pipenv <span style="font-size: 16px;"> </span></td></tr>
    <tr><td style="text-align: left; padding-left: 0px;"
            > Docker Deployment <span style="font-size: 16px;">( Pull Image, Create Image, Run Image to Container )</span></td></tr>
    <tr><td style="text-align: left; padding-left: 0px;"
            > Flask Web Application <span style="font-size: 16px;">( Flask API with a JSON object )</span></td></tr>
  </table>
  </div>

</div>

**Dataset Info**

> https://www.kaggle.com/datasets/kapturovalexander/bank-credit-scoring/data

**About Dataset**

Split dataset into train and test and clean it from null values

**TASK**

**Homework**

In this homework, we will use Bank credit scoring dataset from [here](https://www.kaggle.com/datasets/kapturovalexander/bank-credit-scoring/data).

> **Note**: sometimes your answer doesn't match one of the options exactly. That's fine. 
Select the option that's closest to your solution.

> **Note**: we recommend using python 3.10 in this homework.



**Submit the results**

- Submit your results here: https://forms.gle/gfruq6FGoLass3Ff9
- If your answer doesn't match options exactly, select the closest one.
- You can submit your solution multiple times. In this case, only the last submission will be used

**Deadline**

The deadline for submitting is October 16 (Monday), 23:00 CET. After that the form will be closed.

<div style="letter-spacing:normal; opacity:1.;">
  <h1 style="text-align:center; background-color: lightsalmon; color: Jaguar; border-radius:10px; font-family:monospace; border-radius:20px;
            line-height:1.4; font-size:32px; font-weight:bold; text-transform: uppercase; padding: 9px;">
            <strong>1. Import Libraries & Ingest Data</strong></h1>   
</div>

> ⚠️ Not Recommended conda `base` env, work on `venv`

- https://docs.conda.io/projects/conda/en/4.6.0/_downloads/52a95608c49671267e40c689e0bc00ca/conda-cheatsheet.pdf

**You must use the `--no-deps` option in the pip install command in order to avoid bundling dependencies into your conda-package.**

If you run pip install without the `--no-deps` option, pip will often install dependencies in your conda recipe and those dependencies will become part of your package. <br>
This wastes space in the package and `increases the risk of file overlap`, file clobbering, and broken packages.

There might be cases where you want to install a package directly from a local directory or a specific location, without relying on the package indexes.<br>
In such situations, you can use the `--no-index` option to tell pip not to look for the package in any indexes.

```py
# new conda virtual environment
conda create --name "classifier" python=3.10 jupyter -y
conda activate "classifier"

# if The environment is inconsistent, try below
conda update -n base -c defaults conda --force-reinstall
conda install anaconda --force-reinstall

# get dependencies
pip freeze > requirements.txt
conda list -e > requirements.txt

# install all package dependencies
pip install -r requirements.txt
conda install -c conda-forge pandas==2.0.2 -q -y
```

In [1]:
!python -V

Python 3.10.12


In [2]:
%%writefile requirements.txt 
# To get started with MLflow you'll need to install the appropriate Python package.

# for parquet file
pyarrow
fastparquet
# orjson is a fast, correct JSON library
orjson
numpy
pandas
matplotlib
seaborn
scipy
tqdm
joblib

# ML Model packages
scikit-learn
xgboost
optuna
hyperopt

# Optionally
jupyter
ipykernel
ipywidgets
pipenv

Overwriting requirements.txt


In [3]:
# !pip install jupyter ipykernel ipywidgets
# !pip install pyarrow fastparquet orjson pandas matplotlib seaborn
# !pip install scikit-learn xgboost optuna hyperopt
# !pip install tqdm requests
# !pip install pipenv

In [4]:
import os, sys, platform, IPython.display

# pip install --no-deps --no-index --force-reinstall --no-cache-dir --user
# !{sys.executable} -m pip install -Uq -r requirements.txt --no-cache-dir --user
# !jupyter nbextension enable --py widgetsnbextension

# IPython.display.clear_output()
print("Platform:", platform.system())  # platform.platform()
print("Python  :", platform.python_version())  # sys.version
print("Actv Env:", os.getenv('CONDA_DEFAULT_ENV', 'Not Found Conda Env'))

Platform: Windows
Python  : 3.10.12
Actv Env: lin-reg


In [5]:
import numpy as np
import pandas as pd
import matplotlib as mpl
import matplotlib.pyplot as plt
import seaborn as sns

import statsmodels.api as sm
# import scipy
# import scipy.stats as stats
# from scipy.sparse import hstack
from scipy.stats import skew, norm, gamma, probplot

from sklearn.feature_extraction import DictVectorizer
from sklearn.preprocessing import OneHotEncoder
# from sklearn.preprocessing import StandardScaler, MinMaxScaler, RobustScaler

from sklearn.linear_model import LinearRegression, TweedieRegressor, LogisticRegression
# from sklearn.linear_model import ElasticNet, Lasso, LassoCV, Ridge, RidgeCV
# from sklearn.svm import LinearSVR
# from sklearn.ensemble import RandomForestRegressor
# from sklearn.ensemble import GradientBoostingRegressor, ExtraTreesRegressor

from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score
from sklearn.metrics import mutual_info_score
from sklearn.metrics import make_scorer
from sklearn.metrics import precision_recall_fscore_support, fbeta_score
from sklearn.metrics import confusion_matrix, precision_recall_curve
from sklearn.metrics import roc_curve, auc, roc_auc_score, average_precision_score
# from sklearn.model_selection import GridSearchCV

# import xgboost as xgb
# Hyperparameter Optimization with Tree-structured Parzen Estimator (TPE) algorithm
# from hyperopt import fmin, tpe, hp, STATUS_OK, Trials
# from hyperopt.pyll import scope
import optuna

import os
import sys
import pickle
import zipfile
# import boto3
# import click
# import pathlib
# import argparse
# import requests
# import urllib.request
from glob import glob
# from tqdm import tqdm           # console-based
# from tqdm.notebook import tqdm  # jupyter-based
from tqdm.auto import tqdm        # automatically selects
# tqdm._instances.clear()
# Import joblib for model persistence
# from joblib import load, dump

# memory management performs garbage collection 
import gc
gc.collect()

0

In [6]:
# !pip install session_info
import session_info
session_info.show(html=False)

-----
matplotlib          3.8.1
numpy               1.25.2
optuna              3.3.0
pandas              2.1.0
scipy               1.11.3
seaborn             0.13.0
session_info        1.0.0
sklearn             1.3.2
statsmodels         0.14.0
tqdm                4.66.1
-----
IPython             8.12.0
jupyter_client      7.4.9
jupyter_core        5.3.0
jupyterlab          3.6.3
notebook            6.5.4
-----
Python 3.10.12 | packaged by Anaconda, Inc. | (main, Jul  5 2023, 19:09:20) [MSC v.1916 64 bit (AMD64)]
Windows-10-10.0.22621-SP0
-----
Session information updated at 2023-11-05 05:26


In [7]:
import os
# Get the current working directory
# current_dir = os.getcwd()

# Create a new directory for storing data
os.makedirs('./data', exist_ok=True)
os.makedirs('./pycode', exist_ok=True)
os.makedirs('./output', exist_ok=True)
os.makedirs('./model', exist_ok=True)

<div style="letter-spacing:normal; opacity:1.;">
  <h1 style="text-align:center; background-color: lightsalmon; color: Jaguar; border-radius:10px; font-family:monospace; border-radius:20px;
            line-height:1.4; font-size:32px; font-weight:bold; text-transform: uppercase; padding: 9px;">
            <strong>2. Recognizing and Understanding Data</strong></h1>   
</div>

## Downloading the data

In this homework, we will use Bank credit scoring dataset from [here](https://www.kaggle.com/datasets/kapturovalexander/bank-credit-scoring/data).

### Ingest Data [wget](https://linuxways.net/centos/linux-wget-command-with-examples/) or [curl](https://daniel.haxx.se/blog/2020/09/10/store-the-curl-output-over-there/)

In [8]:
import zipfile

# Read the data for (May 2023)
# Specify the ZIP file path
zip_filename = 'data/archive.zip'
csv_filename = 'bank.csv'
csv_filepath = f'data/{csv_filename}'

# Read the extracted CSV file

# Read Specify the CSV file in the ZIP
with zipfile.ZipFile(zip_filename, 'r') as zip_ref:
    # zip_ref.extract(csv_filename, path='.')
    with zip_ref.open(csv_filename) as f_out:
        df = pd.read_csv(f_out, sep=';')

In [9]:
print("Shape of DataFrame:", df.shape)
df.head()

Shape of DataFrame: (4521, 17)


age          job  marital  education default  balance housing loan  \
0   30   unemployed  married    primary      no     1787      no   no   
1   33     services  married  secondary      no     4789     yes  yes   
2   35   management   single   tertiary      no     1350     yes   no   
3   30   management  married   tertiary      no     1476     yes  yes   
4   59  blue-collar  married  secondary      no        0     yes   no   

    contact  day month  duration  campaign  pdays  previous poutcome   y  
0  cellular   19   oct        79         1     -1         0  unknown  no  
1  cellular   11   may       220         1    339         4  failure  no  
2  cellular   16   apr       185         1    330         1  failure  no  
3   unknown    3   jun       199         4     -1         0  unknown  no  
4   unknown    5   may       226         1     -1         0  unknown  no

## Question 1

* Install Pipenv
* What's the version of pipenv you installed?
* Use `--version` to find out

In [10]:
# !pip install pipenv -Uq
# !pip show pipenv

In [11]:
!pipenv --version

pipenv, version 2023.10.24


#### Answer 1: pipenv, version 2023.10.24

## Question 2

* Use Pipenv to install Scikit-Learn version 1.3.1
* What's the first hash for scikit-learn you get in Pipfile.lock?

> **Note**: you should create an empty folder for homework
and do it there.

```bash
# Create the empty folder
mkdir ml-zoomcamp-module-5-env

# Inside the empty folder, run the following command to initialize a new Pipenv environment:
cd "ml-zoomcamp-module-5-env"

# Create a new virtual environment using desired Python version, Modify Pipfile desired library
pipenv --python 3.10

# Install scikit-learn version 1.3.1:
pipenv install scikit-learn==1.3.1

# check version of scikit-learn installed within Pipenv environment
pipenv run pip show scikit-learn
```

Sample Pipfle
```pipfile
[[source]]
url = "https://pypi.org/simple"
verify_ssl = true
name = "pypi"

[packages]
scikit-learn = "==1.3.1"
flask = "*"
waitress = "*"

[dev-packages]

[requires]
python_version = "3.10"
```

## Create and Check Pipenv environment

In [20]:
# mkdir ml-zoomcamp-module-5-env
# cd "ml-zoomcamp-module-5-env"
# pipenv --python 3.10
# pipenv install scikit-learn==1.3.1
!cd "ml-zoomcamp-module-5-env" & pipenv run pip show scikit-learn

Name: scikit-learn
Version: 1.3.1
Summary: A set of python modules for machine learning and data mining
Home-page: http://scikit-learn.org
Author: 
Author-email: 
License: new BSD
Location: c:\users\clk\.virtualenvs\ml-zoomcamp-module-5-env-asuwqmac\lib\site-packages
Requires: joblib, numpy, scipy, threadpoolctl
Required-by: 


In [23]:
import json

# Specify the path to your Pipfile.lock
pipfile_lock_path = 'ml-zoomcamp-module-5-env\Pipfile.lock'

# Read the Pipfile.lock
with open(pipfile_lock_path, 'r') as lock_file:
    lock_data = json.load(lock_file)

# Get the first hash for scikit-learn
scikit_learn_hash = lock_data['default']['scikit-learn']['hashes'][0]

# Print the first hash
print("The first hash for scikit-learn is:", scikit_learn_hash)

The first hash for scikit-learn is: sha256:0c275a06c5190c5ce00af0acbb61c06374087949f643ef32d355ece12c4db043


#### Answer 2: The first hash for scikit-learn is: sha256:0c275a06c5190c5ce00af0acbb61c06374087949f643ef32d355ece12c4db043

# Prepare Model and Preprocessor

We've prepared a dictionary vectorizer and a model.

They were trained (roughly) using this code:

```python
features = ['job','duration', 'poutcome']
dicts = df[features].to_dict(orient='records')

dv = DictVectorizer(sparse=False)
X = dv.fit_transform(dicts)

model = LogisticRegression().fit(X, y)
```

> **Note**: You don't need to train the model. This code is just for your reference.

And then saved with Pickle. Download them:

* [DictVectorizer](https://github.com/DataTalksClub/machine-learning-zoomcamp/tree/master/cohorts/2023/05-deployment/homework/dv.bin?raw=true)
* [LogisticRegression](https://github.com/DataTalksClub/machine-learning-zoomcamp/tree/master/cohorts/2023/05-deployment/homework/model1.bin?raw=true)

With `wget`:

```bash
PREFIX=https://raw.githubusercontent.com/DataTalksClub/machine-learning-zoomcamp/master/cohorts/2023/05-deployment/homework
wget $PREFIX/model1.bin
wget $PREFIX/dv.bin
```

In [24]:
PREFIX = "https://raw.githubusercontent.com/DataTalksClub/machine-learning-zoomcamp/master/cohorts/2023/05-deployment/homework"
!curl -Os --output-dir "model" $PREFIX/model1.bin
!curl -Os --output-dir "model" $PREFIX/dv.bin

## Question 3

Let's use these models!

* Write a script for loading these models with pickle
* Score this client:

```json
{"job": "retired", "duration": 445, "poutcome": "success"}
```

What's the probability that this client will get a credit? 

* 0.162
* 0.392
* 0.652
* 0.902

If you're getting errors when unpickling the files, check their checksum:

```bash
$ md5sum model1.bin dv.bin
8ebfdf20010cfc7f545c43e3b52fc8a1  model1.bin
924b496a89148b422c74a62dbc92a4fb  dv.bin
```

In [25]:
!CertUtil -hashfile model/model1.bin MD5
!CertUtil -hashfile model/dv.bin MD5

MD5 hash of model/model1.bin:
8ebfdf20010cfc7f545c43e3b52fc8a1
CertUtil: -hashfile command completed successfully.
MD5 hash of model/dv.bin:
924b496a89148b422c74a62dbc92a4fb
CertUtil: -hashfile command completed successfully.


### q3_test

In [27]:
import pickle

def load(filename: str):
    # Load the object with pickle
    with open(filename, 'rb') as f_in:
        return pickle.load(f_in)
    
# Load the data preprocessing object and the trained model
dv, model = load('model/dv.bin'), load('model/model1.bin')

In [28]:
# Define the client data as a dictionary
client_data = {
    "job": "retired",
    "duration": 445,
    "poutcome": "success"
}

# Use the data preprocessing object to transform the client data
X = dv.transform([client_data])

# Make a probability prediction using the loaded model
probability = model.predict_proba(X)[:, 1]

print(f"The probability that this client will get a credit is: {probability[0]:.3f}")

The probability that this client will get a credit is: 0.902


### Define q3_test

In [102]:
%%writefile pycode/q3_test.py

import pickle

def load(filename: str):
    # Load the object with pickle
    with open(filename, 'rb') as f_in:
        return pickle.load(f_in)
    
# Load the data preprocessing object and the trained model
dv, model = load('../model/dv.bin'), load('../model/model1.bin')

# Define the client data as a dictionary
client_data = {
    "job": "retired",
    "duration": 445,
    "poutcome": "success"
}

# Use the data preprocessing object to transform the client data
X = dv.transform([client_data])

# Make a probability prediction using the loaded model
probability = model.predict_proba(X)[:, 1]

print(f"The probability that this client will get a credit is: {probability[0]:.3f}")

Overwriting pycode/q3_test.py


In [30]:
!cd "ml-zoomcamp-module-5-env" & pipenv run python "../pycode/q3_test.py"

The probability that this client will get a credit is: 0.902


What's the probability that this client will get a credit? 

* 0.162
* 0.392
* 0.652
* **0.902**

# Write Flask Code

## Question 4

Now let's serve this model as a web service

* Install Flask and gunicorn (or waitress, if you're on Windows)
* Write Flask code for serving the model
* Now score this client using `requests`:

```python
url = "YOUR_URL"
client = {"job": "unknown", "duration": 270, "poutcome": "failure"}
requests.post(url, json=client).json()
```

What's the probability that this client will get a credit?

* 0.140
* 0.440
* 0.645
* 0.845


```bash
# Install Flask and gunicorn (or waitress for Windows):
pipenv install flask
pipenv install gunicorn
pipenv install waitress
```

### Define q4_predictor

In [103]:
%%writefile pycode/q4_predictor.py

from flask import Flask, request, jsonify
import pickle


def load(filename: str):
    # Load the object with pickle
    with open(filename, 'rb') as f_in:
        return pickle.load(f_in)

# Load the data preprocessing object and the trained model
dv, model = load('../model/dv.bin'), load('../model/model1.bin')

app = Flask(__name__)

@app.route('/predict', methods=['POST'])
def predict():
    # client = request.get_json()
    client = request.json

    # Use the data preprocessing object to transform the input data
    X = dv.transform([client])

    # Make a probability prediction using the loaded model
    probability = model.predict_proba(X)[:, 1]

    result = {
        'get_credit_probability': float(probability),
        'get_credit': bool(probability >= 0.5)
    }

    return jsonify(result)

if __name__ == '__main__':
    app.run(debug=True, host='0.0.0.0', port=9696, use_reloader=False)

Overwriting pycode/q4_predictor.py


### Run the Flask App using Pipenv

```py
# activate pipenv via Using Pipfile, cd env_path
pip shell
```

In [32]:
# cd app.py_path

# Run the Flask app using Pipenv, which ensures it runs within the virtual environment
# gunicorn -w 4 -b 0.0.0.0:8000 app:app  # For Linux/macOS

# if you're on Windows and using Waitress
# waitress-serve --port=8000 app:app

## Score the client using requests

### Define q4_test

In [100]:
%%writefile pycode/q4_test.py

import requests

# To access your Flask application served by Waitress, 
# you should use http://localhost:8000 or http://127.0.0.1:8000 in your web browser. 
url = "http://localhost:9696/predict"  # Replace with your actual URL

client = {"job": "unknown", "duration": 270, "poutcome": "failure"}

response = requests.post(url, json=client).json()
print(f"Does the client approve the loan?: {response['get_credit']}")
print(f"The probability that this client will get a credit is: {response['get_credit_probability']:.3f}")

Writing pycode/q4_test.py


In [61]:
!python pycode/q4_test.py

Does the client approve the loan?: False
The probability that this client will get a credit is: 0.140


What's the probability that this client will get a credit?

* **0.140**
* 0.440
* 0.645
* 0.845

# Docker

Install [Docker](https://github.com/DataTalksClub/machine-learning-zoomcamp/blob/master/05-deployment/06-docker.md).<br>
We will use it for the next two questions.

For these questions, we prepared a base image: `svizor/zoomcamp-model:3.10.12-slim`.<br>
You'll need to use it (see Question 5 for an example).

```sh 
# Pull the Docker image:
docker pull svizor/zoomcamp-model:3.10.12-slim

# Check the size of the downloaded image:
docker images
```

This image is based on `python:3.10.12-slim` and has a logistic regression model<br>
(a different one) as well a dictionary vectorizer inside. 

This is how the Dockerfile for this image looks like:

```docker 
FROM python:3.10.12-slim
WORKDIR /app
COPY ["model2.bin", "dv.bin", "./"]
```

We already built it and then pushed it to [`svizor/zoomcamp-model:3.10.12-slim`](https://hub.docker.com/r/svizor/zoomcamp-model).

> **Note**: You don't need to build this docker image, it's just for your reference.

## Question 5

Download the base image `svizor/zoomcamp-model:3.10.12-slim`. You can easily make it by using [docker pull](https://docs.docker.com/engine/reference/commandline/pull/) command.

So what's the size of this base image?

* 47 MB
* 147 MB
* 374 MB
* 574 MB

You can get this information when running `docker images` - it'll be in the "SIZE" column.

### Download Docker Image

In [34]:
!docker pull svizor/zoomcamp-model:3.10.12-slim

3.10.12-slim: Pulling from svizor/zoomcamp-model
52d2b7f179e3: Pulling fs layer
2b8a9a2240c1: Pulling fs layer
618a49bbc6c6: Pulling fs layer
708d13076071: Pulling fs layer
90b76fc2ebde: Pulling fs layer
a7c599e7b037: Pulling fs layer
faa7e3220190: Pulling fs layer
708d13076071: Waiting
a7c599e7b037: Waiting
faa7e3220190: Waiting
90b76fc2ebde: Waiting
2b8a9a2240c1: Download complete
708d13076071: Download complete
90b76fc2ebde: Verifying Checksum
90b76fc2ebde: Download complete
618a49bbc6c6: Verifying Checksum
618a49bbc6c6: Download complete
faa7e3220190: Download complete
a7c599e7b037: Download complete
52d2b7f179e3: Verifying Checksum
52d2b7f179e3: Download complete
52d2b7f179e3: Pull complete
2b8a9a2240c1: Pull complete
618a49bbc6c6: Pull complete
708d13076071: Pull complete
90b76fc2ebde: Pull complete
a7c599e7b037: Pull complete
faa7e3220190: Pull complete
Digest: sha256:e8441100b9d8da56344f50c673eb2daded3c61ce9565e45c3592c02f34fb3149
Status: Downloaded newer image for svizor/zoomc


What's Next?
  View a summary of image vulnerabilities and recommendations → docker scout quickview svizor/zoomcamp-model:3.10.12-slim


In [35]:
!docker images

REPOSITORY              TAG            IMAGE ID       CREATED       SIZE
svizor/zoomcamp-model   3.10.12-slim   08266c8f0c4b   3 weeks ago   147MB


## Question 6

Let's run your docker container!

After running it, score this client once again:

```python
url = "YOUR_URL"
client = {"job": "retired", "duration": 445, "poutcome": "success"}
requests.post(url, json=client).json()
```

What's the probability that this client will get a credit now?

* 0.168
* 0.530
* 0.730
* 0.968

### Define q6_predictor

In [60]:
%%writefile pycode/q6_predictor.py

from flask import Flask, request, jsonify
import pickle


def load(filename: str):
    # Load the object with pickle
    with open(filename, 'rb') as f_in:
        return pickle.load(f_in)

# Load the data preprocessing object and the trained model
dv, model = load('dv.bin'), load('model2.bin')

app = Flask(__name__)

@app.route('/predict', methods=['POST'])
def predict():
    # client = request.get_json()
    client = request.json

    # Use the data preprocessing object to transform the input data
    X = dv.transform([client])

    # Make a probability prediction using the loaded model
    probability = model.predict_proba(X)[:, 1]

    result = {
        'get_credit_probability': float(probability),
        'get_credit': bool(probability >= 0.5)
    }

    return jsonify(result)

if __name__ == '__main__':
    app.run(debug=True, host='0.0.0.0', port=9696, use_reloader=False)

Overwriting pycode/q6_predictor.py


## Dockerfile, Modify Docker Image

In [85]:
%%writefile Dockerfile

# Use the ml-zoomcamp image
FROM svizor/zoomcamp-model:3.10.12-slim

# Set the working directory in the container
WORKDIR /app

# Copy the application code into the container
# Copy the Pipfile and Pipfile.lock into the container
COPY ["pycode/", "ml-zoomcamp-module-5-env/", "./"]

# Install pipenv and the project dependencies
RUN pip install pipenv
RUN pipenv install --system --deploy --ignore-pipfile

# Expose the port that Flask will run on
EXPOSE 9696

# Start the Flask application using pipenv
# CMD ["pipenv", "run", "waitress-serve", "--listen=0.0.0.0:9696", "q6_predictor:app"]
ENTRYPOINT ["pipenv", "run", "waitress-serve", "--listen=0.0.0.0:9696", "q6_predictor:app"]

Overwriting Dockerfile


In [86]:
!docker build -t my-flask-app .

#0 building with "default" instance using docker driver

#1 [internal] load .dockerignore
#1 transferring context: 2B done
#1 DONE 0.0s

#2 [internal] load build definition from Dockerfile
#2 transferring dockerfile: 744B 0.0s done
#2 DONE 0.0s

#3 [internal] load metadata for docker.io/svizor/zoomcamp-model:3.10.12-slim
#3 DONE 0.0s

#4 [1/5] FROM docker.io/svizor/zoomcamp-model:3.10.12-slim
#4 DONE 0.0s

#5 [internal] load build context
#5 transferring context: 345B done
#5 DONE 0.0s

#6 [2/5] WORKDIR /app
#6 CACHED

#7 [4/5] RUN pip install pipenv
#7 CACHED

#8 [3/5] COPY [pycode/, ml-zoomcamp-module-5-env/, ./]
#8 CACHED

#9 [5/5] RUN pipenv install --system --deploy --ignore-pipfile
#9 CACHED

#10 exporting to image
#10 exporting layers done
#10 writing image sha256:2a0148db78e53b8db957d0c2c192bf5773a0a4f811f5780ec7f7fce9b60e1dc4 done
#10 naming to docker.io/library/my-flask-app done
#10 DONE 0.0s

What's Next?
  View a summary of image vulnerabilities and recommendations → docker

In [87]:
!docker images

REPOSITORY              TAG            IMAGE ID       CREATED          SIZE
my-flask-app            latest         2a0148db78e5   10 minutes ago   505MB
svizor/zoomcamp-model   3.10.12-slim   08266c8f0c4b   3 weeks ago      147MB


### Run Container

In [99]:
# Run Container
# -d  means container in detached mode (in the background)
# -it means container in interactively
# -rm means Automatically removes the container when it stops.
# -p 9696:9696: Maps port 9696 on the host to port 9696 in the container.
# docker run -d --rm -p 9696:9696 --name my-flask-container my-flask-app

In [89]:
# List running containers and their ports
!docker ps -a

CONTAINER ID   IMAGE          COMMAND                  CREATED          STATUS          PORTS                    NAMES
0a5893d7b18f   my-flask-app   "pipenv run waitress…"   19 seconds ago   Up 18 seconds   0.0.0.0:9696->9696/tcp   my-flask-container


In [98]:
# !sudo netstat -tuln | grep 8000
# !netstat -ano | findstr "9696"
# !taskkill /F /PID xxxx

## Score the client using requests

### Define q6_test

In [101]:
%%writefile pycode/q6_test.py

import requests

url = "http://localhost:9696/predict"  # Replace with your actual URL
client = {"job": "retired", "duration": 445, "poutcome": "success"}

response = requests.post(url, json=client).json()
print(f"Does the client approve the loan?: {response['get_credit']}")
print(f"The probability that this client will get a credit is: {response['get_credit_probability']:.3f}")

Writing pycode/q6_test.py


In [96]:
!python pycode/q6_test.py

Does the client approve the loan?: True
The probability that this client will get a credit is: 0.727


### Stop Container

In [97]:
# Stop container
!docker stop my-flask-container

my-flask-container


What's the probability that this client will get a credit now?

* 0.168
* 0.530
* **0.730**
* 0.968

## End of Project